### PR TITLE
Rewrite subsidy limit tests (1.21)

### PR DESCRIPTION
**Pull request by rnicoll**
Rebuild subsidy limit tests:

* Subsidy values are deterministic even if pseudo-random, so test against the expected value rather than a range.
* Break block heights into ranges so total fits into `CAmount` rather than overflows.
* Use previous susidy values to vary the previous hash value on each calculation, rather than always using zero.

### CODE REVIEW

**Comment by patricklodder** [comment]
closing & reopening to re-trigger Cirrus CI

**Comment by patricklodder** [comment]
There's still one job broken @ cirrus: https://cirrus-ci.com/task/5353022818615296 - I've restarted it a couple of times, but doesn't seem to want to succeed. Timeout?

**Comment by patricklodder** [comment]
The main issue blocking progress on this now is 2b82175 because it COULD trigger a chaintip fork as consensus is built over validity of scrypt digests. We either need a test that proves that said overflow is not an issue, or a consensus plan with a softfork trigger (meaning a version-conditional implementation of this code) if this were to make it in.

**Comment by rnicoll** [comment]
> The main issue blocking progress on this now is [2b82175](https://github.com/dogecoin/dogecoin/commit/2b821751fe1b2ed9fa350ada9100a86c986278b4) because it COULD trigger a chaintip fork as consensus is built over validity of scrypt digests. We either need a test that proves that said overflow is not an issue, or a consensus plan with a softfork trigger (meaning a version-conditional implementation of this code) if this were to make it in.

That commit is in 1.14 as well, although notably it's something we differ from Litecoin on. See https://github.com/dogecoin/dogecoin/commit/57dae15a77410684126a9dfbcc6a72faff827906

**Comment by patricklodder** [comment]
Oof! Missed that! Do you have any documentation that triggered the change in 1.14? Would be awesome to have a test vector.

I'll proceed to review this.

File: src/miner.cpp [hunk 1]
```diff
@@ -72,7 +72,7 @@ static BlockAssembler::Options DefaultOptions()
     BlockAssembler::Options options;
     options.nBlockMaxWeight = gArgs.GetArg("-blockmaxweight", DEFAULT_BLOCK_MAX_WEIGHT);
     CAmount n = 0;
-    if (gArgs.IsArgSet("-blockmintxfee") && ParseMoney(gArgs.GetArg("-blockmintxfee", ""), n)) {
+    if (gArgs.IsArgSet("-blockmintxfee") && ParseMoney(gArgs.GetArg("-blockmintxfee", ""), n) && n <= MAX_FEE_RATE) {
```
Comment (patricklodder) [inline src/miner.cpp:75]: Can we print a warning to log here? All the other places are throwing fatals, so would be good to at least have a warning to document unexpected behavior?

**Review by patricklodder — CHANGES_REQUESTED** [review]
General: needs rebase to remove the cherry-pick and revert.

**Comment by rnicoll** [comment]
> 
> 
> Oof! Missed that! Do you have any documentation that triggered the change in 1.14? Would be awesome to have a test vector.
> 
> I'll proceed to review this.

If I recall correctly (it was a while back), Bitcoin Core replaced OpenSSL's SHA256 so they knew it couldn't change, so I moved Dogecoin over to their implementation for consistency.

File: ci/test/00_setup_env_native_asan.sh [hunk 2]
```diff
@@ -7,8 +7,11 @@
 export LC_ALL=C.UTF-8
 
 export CONTAINER_NAME=ci_native_asan
-export PACKAGES="clang llvm python3-zmq qtbase5-dev qttools5-dev-tools libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libqrencode-dev libsqlite3-dev"
-export DOCKER_NAME_TAG=ubuntu:20.04
+export PACKAGES="clang llvm python3-zmq qtbase5-dev qttools5-dev-tools libevent-dev bsdmainutils libboost-dev libboost-system-dev libboost-filesystem-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libnatpmp-dev libzmq3-dev libqrencode-dev libsqlite3-dev"
+export DOCKER_NAME_TAG=ubuntu:hirsute
```
[the remark this answers is not in the document]
Comment (rnicoll) [reply ci/test/00_setup_env_native_asan.sh:11]: It's from a failed attempt to get the test not to hang, I'll revert.

File: src/miner.cpp (hunk 1 above)
Comment (rnicoll) [reply src/miner.cpp]:
```
2021-07-11T22:16:31Z DefaultOptions(): Specified -blockmintxfee is either not a valid amount, or is greater than the maximum fee rate 2100000000000000.
```
Done and verified.

**Review by patricklodder — APPROVED** [review]
utACK
